### ISSUE TITLE
Unexpected @extend behavior with > selector

### ISSUE THREAD

**Issue by jedfoster** [sass/sass#657]
I'm seeing some unexpected behavior with an @extended placeholder class containing >.

```
%foo
  > header
    hgroup
      font-size: 3rem

#unexpected
  @extend %foo
  margin: 1em
  > section
    @extend %foo
```

This is the output:

```
#unexpected > header hgroup {
  font-size: 3rem;
}

#unexpected {
  margin: 1em;
}
```

This is the expected output:

```
#expected > header hgroup, #expected > section > header hgroup {
  font-size: 3rem; 
}

#expected {
  margin: 1em; 
}
```

See it here: http://sassmeister.com/gist/4760254

However the same set of rules with + where the > is produces the expected chain of selectors.  See it here: http://sassmeister.com/gist/4760268

I can't understand why Sass is doing this. Any ideas?

**Comment by jedfoster** [comment]
The links above are broken because that's what happens when I delete Gists....

Here's a new link, with >: http://sassmeister.com/gist/4988878

Here's a new link with +: http://sassmeister.com/gist/4988881

**Comment by nex3** [comment]
Upon investigation, I've realized that this is actually the intended behavior--moreover, it will do what you want it to. `#expected > header hgroup` is a superselector of `#expected > section > header hgroup`. Every element that's matched by `#expected > section > header hgroup` is also matched by `#expected > header hgroup`. Sass detects this and helpfully removes the unnecessary selector.

**Comment by jedfoster** [comment]
I agree that `#expected > header hgroup` is a super selector, however what I need is the specificity of `#expected > section > header hgroup`. I have a particular set of styles that I need to apply to `hgroup`s inside `header`s that are immediate descendants of a `section` which is an immediate descendant of `#expected`, and which styles override some set by `#expected > header hgroup`. 

To state it another way, yes every element in `#expected > section > header hgroup` is also in `#expected > header hgroup`, but not every element in the latter matches the former. I'd post a Venn diagram, but I'm on my iPhone on the train home, so just imagine two concentric circles with the `section` selector in the center. 

This seems to me to be an over-optimization on Sass' part. There are things I want Sass to optimize for me, but not this.

**Comment by nex3** [comment]
Sass provides a very specific set of guarantees about the specificity of selectors it generates via `@extend`. Due to Sass's nature as a preprocessor, it's impossible to be entirely consistent with user expectations here, but the guarantees we do offer are the best we can do to make it easy to reason about specificity when using `@extend`. The pertinent guarantee is what I call "the second law of extend": **the specificity of a generated selector should never be less than the specificity of the extending selector**.

In this case, the generated selector is `#expected > header hgroup` with specificity 102, and the extending selector is `#unexpected > section` with specificity 101. This follows the second law of extend, and so is correct.

For further reference, see [this comment](https://github.com/nex3/sass/issues/324#issuecomment-4607184) where I expound at length the details of Sass's handling of specificity.

### tl;dr

Sass does the best it can to handle specificity consistently. If you want more consistency, use `!important`.

**Comment by jedfoster** [comment]
Hmm. I calculate `#expected > section > header hgroup` as having a specificity of 103, not 101. 

Also, as demonstrated here, http://sassmeister.com/gist/5035973, if I change `> section` to just `section`, I get the expected CSS, so it seems there's something about the `>` selector.

That said, this seems to be an extreme corner case. I've coded around this issue in my project. Until someone else runs into this, I think we can let it stand.

Thank you for all you hard work on Sass!

**Comment by nex3** [comment]
> Hmm. I calculate `#expected > section > header hgroup` as having a specificity of 103, not 101. 

The specificity of `#expected > section > header hgroup` doesn't matter. The only comparison we're making is between `#unexpected > section` (the extending selector) and `#expected > header hgroup` (the simplest generated selector). The only specificity guarantee Sass gives is that the generated selector will have greater than or equal specificity to the extending selector, which it does.

> Also, as demonstrated here, http://sassmeister.com/gist/5035973, if I change `> section` to just `section`, I get the expected CSS, so it seems there's something about the `>` selector.

This is because `#expected > header hgroup` is no longer a superselector of the newly-generated selector (`#expected section > header hgroup`).
